.github/scripts/generate_ci_workflows.py

```diff
@@ -1,123 +1,106 @@
 
-def PyTorchLinuxWorkflow(
-    *,
-    build_environment: str,
-    docker_image_base: str,
-    test_runner_type: str,
-    on_pull_request: bool = False,
-    enable_doc_jobs: bool = False,
-    enable_jit_legacy_test: YamlShellBool = "''",
-    enable_multigpu_test: YamlShellBool = "''",
-    enable_nogpu_no_avx_test: YamlShellBool = "''",
-    enable_nogpu_no_avx2_test: YamlShellBool = "''",
-    enable_slow_test: YamlShellBool = "''",
-    num_test_shards: int = 1,
-    is_scheduled: Optional[str] = None,
-    is_libtorch: bool = False,
-    exclude_test: bool = False,
-) -> PyTorchWorkflow:
-    return {
-        "build_environment": build_environment,
-        "docker_image_base": docker_image_base,
-        "test_runner_type": test_runner_type,
-        "on_pull_request": on_pull_request,
-        "is_scheduled": is_scheduled,
-        "enable_doc_jobs": enable_doc_jobs,
-        "enable_jit_legacy_test": enable_jit_legacy_test,
-        "enable_multigpu_test": enable_multigpu_test,
-        "enable_nogpu_no_avx_test": enable_nogpu_no_avx_test,
-        "enable_nogpu_no_avx2_test": enable_nogpu_no_avx2_test,
-        "enable_slow_test": enable_slow_test,
-        "num_test_shards": num_test_shards,
-        "is_libtorch": is_libtorch,
-        "exclude_test": is_libtorch or exclude_test,   # libtorch is build only
-    }
+    def __post_init__(self) -> None:
+        if self.is_libtorch:
+            self.exclude_test = True
+        self.only_build_on_pull_request = self.only_build_on_pull_request and self.on_pull_request
 
+    def assert_valid(self) -> None:
+        err_message = f"invalid test_runner_type for {self.arch}: {self.test_runner_type}"
+        if self.arch == 'linux':
+            assert self.test_runner_type in LINUX_RUNNERS, err_message
+        if self.arch == 'windows':
+            assert self.test_runner_type in WINDOWS_RUNNERS, err_message
 
-def generate_workflow_file(
-    *,
-    workflow: PyTorchWorkflow,
-    workflow_template: jinja2.Template,
-) -> Path:
-    output_file_path = GITHUB_DIR / f"workflows/{workflow['build_environment']}.yml"
-    with open(output_file_path, "w") as output_file:
-        GENERATED = "generated"
-        output_file.writelines([f"# @{GENERATED} DO NOT EDIT MANUALLY\n"])
-        output_file.write(workflow_template.render(**workflow))
-        output_file.write("\n")
-    return output_file_path
+    def generate_workflow_file(self, workflow_template: jinja2.Template) -> None:
+        output_file_path = GITHUB_DIR / f"workflows/{workflow.build_environment}.yml"
+        with open(output_file_path, "w") as output_file:
+            GENERATED = "generated"
+            output_file.writelines([f"# @{GENERATED} DO NOT EDIT MANUALLY\n"])
+            output_file.write(workflow_template.render(**asdict(workflow)))
+            output_file.write("\n")
+        print(output_file_path)
 
 
 WINDOWS_WORKFLOWS = [
-    PyTorchWindowsWorkflow(
+    CIWorkflow(
+        arch="windows",
```

Comment:
Should we just make this value a constant? To avoid people mistyping this?

Reply:
this is solved by the `Arch = Literal["windows", "linux"]` definition; see https://github.com/pytorch/pytorch/pull/61879#discussion_r673427592